### DISCUSSION TITLE
Server-side streaming support clarification

### DISCUSSION THREAD

**Discussion by tsegismont** [grpc/grpc-web#1409] [General]
Hi,

In different parts of the documentation, it is said that server-side streaming is only supported  when `grpcwebtext` mode is used. What's the limitation exactly?

I'm asking because I'm currently working on an implementation of [gRPC-Web for Vert.x](https://github.com/eclipse-vertx/vertx-grpc/pull/89) and, during my testing, it seemed to work[1] in both modes.

Do you believe I did something wrong or is it not officially supported because of some limitations in complex cases?

Thank you,
Thomas

[1] Worked with the browser client + Envoy Proxy + Vertx gRPC backend, and browser client + Vertx gRPC-Web backend as well.

**Comment by sampajano** [comment]
@tsegismont Thanks so much for starting the discussion! It's a worthy discussion!

I checked with the original author of the library (@stanley-cheung). He mentioned that at the time of design (pre 2020), binary payload didn't work across browsers when [`Transfer-Encoding`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding) is set to `chunked` (which is what's used with server streaming RPCs).

Do you/we have reason to believe this is still the case, or it has changed since then?

**Reply by tsegismont** [reply]
If I understand correctly, the limitation doesn't come from the client or proxy implementation, but from lack of support from some browsers?

**Reply by sampajano** [reply]
Yes, that's how i've understood the statement.. 

Although i have not researched into the original rationale, nor have i verified in newer browser(s), so i cannot say how true it is..

If i have to guess, there's probably some grounds for the original rationales. So if we can find some original source(s), and reasons to believe that things have evolved since then, i'd have much higher confidence 😃

**Reply by tsegismont** [reply]
Agreed. If you can get a list of tested browsers from the original authors, that would be great. In this case, I would volunteer for a PR that updates the doc (and then users could chose to use server-side streaming with binary format if their website doesn't target the unsupported browsers).

**Reply by sampajano** [reply]
@tsegismont Thanks so much for offering for the change! Appreciate it! 😃

---

Although, i just did a quick test using the [echo example](https://github.com/grpc/grpc-web?tab=readme-ov-file#advanced-demo-browser-echo-app), with the only change to the following config to use `grpcweb` instead of `grpcwebtext`:

https://github.com/grpc/grpc-web/blob/a639b4cf2611de2b68883571787083b73cf61f5e/net/grpc/gateway/docker/commonjs_client/Dockerfile#L19-L21

And it seems that the whole stream body was received it at once..

I'm not sure yet if this is a limitation of our Envoy proxy or the JS client (seems like it shouldn't be the case since it's working for you?), but it doesn't appear to be working out of the box..

I need more debugging to understand what's truly going on, and verify if it's a browser/proxy limitation, or simply a bug, etc..

**Comment by fawdlstty** [comment]
Hello. I used Flutter's GRPC library and compiled it into a web client. After testing, I found that the streaming interface can be called normally, and the protocol content is in binary format. The following is a screenshot of the Chrome browser console:
![captrue](https://github.com/grpc/grpc-web/assets/7323999/7e2fdfe7-b2d4-4149-987d-1ed35d5775c1)

**Reply by fawdlstty** [reply]
“响应标头” means “Response Headers”，and the “请求标头” means “Request Headers”

**Reply by sampajano** [reply]
@fawdlstty Hi! Thanks for the feedback!

If you're using Flutter, chances are you're running grpc-dart underneath:
https://github.com/grpc/grpc-dart

Could you verify?

---

> I found that the streaming interface can be called normally, and the protocol content is in binary format.

By this, do you mean streaming works as expected (e.g. response coming in chunks), despite using binary format?

**Reply by fawdlstty** [reply]
> Could you verify?

It's this repo.
package: https://pub-web.flutter-io.cn/packages/grpc
version: 3.2.3

> By this, do you mean streaming works as expected (e.g. response coming in chunks), despite using binary format?

Yes, I confirm

**Reply by sampajano** [reply]
> > Could you verify?
> 
> Is this repo. package: https://pub-web.flutter-io.cn/packages/grpc version: 3.2.3

Yes i believe so! So that's a project mirroring the implementation in this project, but is done separately at https://github.com/grpc/grpc-dart.

> 
> > By this, do you mean streaming works as expected (e.g. response coming in chunks), despite using binary format?
> 
> Yes, I confirm

Oh wow! Very good to note! That means there would be potential for that to be supported here in grpc-web too!  😃 

Good to note!

---

Do you have a need to use https://github.com/grpc/grpc-web instead of the Dart version?

**Reply by fawdlstty** [reply]
> Do you have a need to use https://github.com/grpc/grpc-web instead of the Dart version?

I want to know the cause of the problem, which has little to do with the actual project, as I can use grpcwebtext to make the project work as expected

**Reply by fawdlstty** [reply]
https://github.com/grpc/grpc-dart/issues/511
It seems to be related to this

**Reply by sampajano** [reply]
> > Do you have a need to use https://github.com/grpc/grpc-web instead of the Dart version?
> 
> I want to know the cause of the problem, which has little to do with the actual project, as I can use grpcwebtext to make the project work as expected

@fawdlstty Ahh that makes sense! Thanks for trying to understand the deeper causes too!

I'm more curious than before as well!

I still need to investigate streaming but yet to have too much time to dig in yet! (As i'm recently busy with Typescript migration).

Streaming is my next focus as i finish with that!
